### PR TITLE
Peering hosts

### PR DESCRIPTION
**Pull request by jcorbin**
First steps towards peering changes:
- refactors the channel peering structure to:
  - by an explicit map-like
  - have explicit peer objects
- adds skeleton peer state objects, only used for selection in static ways presently
- adds support for hosts option
  - [ ] TODO document
  - [x] TODO tests
  - [ ] TODO example usage

reviewers: @kriskowal @Raynos

### CODE REVIEW

**Comment by jcorbin** [comment]
ready for review @kriskowal / @Raynos I want to get this in, then start deving sub channel support by integration with autobahn

**Comment by kriskowal** [comment]
:+1:

File: node/index.js [hunk 1]
```diff
@@ -867,4 +759,259 @@ function TChannelClientOp(req, start) {
...
+    if (threshold === undefined) threshold = self.options.peerScoreThreshold;
+    if (threshold === undefined) threshold = 0;
+
+    var selectedPeer = null, selectedScore = 0;
+    for (var i = 0; i < hosts.length; i++) {
+        var peer = self.add(hosts[i]);
+        var score = peer.state.shouldRequest(options, op);
+        var want = score > threshold &&
+                   (selectedPeer === null || score > selectedScore);
+        self.logger.debug('choose peer score', {
```
Comment (Raynos) [inline node/index.js]: This happens for every `request()`; consider using `.trace()`

File: node/index.js [hunk 2]
```diff
@@ -867,4 +759,259 @@ function TChannelClientOp(req, start) {
...
+    var self = this;
+    self.channel = channel;
+    self.name = name;
+}
+
+TChannelPeerState.prototype.shouldRequest = function shouldRequest(/* options, op */) {
+    // TODO: op isn't quite right currently as a "TChannelClientOp", the
+    // intention is that the other (non-options) arg encapsulates all requests
+    // across retries and setries
+    return 0;
```
Comment (Raynos) [inline node/index.js]: What does `0` means ? I would guess that `0` means that it should never request.

**Comment by Raynos** [comment]
#LEGIT

**Comment by jcorbin** [comment]
Updated:
- renamed / implemented more map-like keys/values/entries/delete
- added addPeer, needed in upcoming sub-channel work
- commented out the score debugging for now, I don't think even trace is right

**Comment by Raynos** [comment]
lgtm :D

**Pull request merged**